src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8376,280 @@ private StatementSyntax ParseLocalDeclarationStatement()
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionTypeIdPart()
+        {
+            var type = ParseType(parentIsParameter: false);
+            var identifier = ParseIdentifierToken();
+
+            var declarator = _syntaxFactory.VariableDeclarator(identifier, null, null);
+            return _syntaxFactory.VariableDeclaration(
+                                    type: type,
+                                    variables: new SeparatedSyntaxList<VariableDeclaratorSyntax>(new SyntaxList<CSharpSyntaxNode>(declarator)),
+                                    deconstruction: null);
+        }
+
+        /// <summary>
+        /// Check ahead for a deconstruction declaration. This requires at least one good-looking variable and the presence of an equals sign.
+        /// Doesn't move the cursor.
+        /// PROTOTYPE(tuples) Can this be done without allocations?
+        /// </summary>
+        private bool IsPossibleDeconstructionDeclaration()
+        {
+            if ((this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken) ||
+                this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                var resetPoint = this.GetResetPoint();
+                try
+                {
+                    var variables = ParseDeconstructionDeclaration(withEquals: false);
+                    var equalsToken = this.EatToken(SyntaxKind.EqualsToken);
+
+                    // We need the equals token and one other confirmation that this is a deconstruction syntax
+                    return DeconstructionVariableLooksGood(variables) && !equalsToken.IsMissing;
+                }
+                finally
+                {
+                    this.Reset(ref resetPoint);
+                    this.Release(ref resetPoint);
+                }
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if one distinct clue is found that this is intended as deconstruction variables
+        /// </summary>
+        private static bool DeconstructionVariableLooksGood(VariableDeclarationSyntax node, bool topLevel = true)
+        {
+            if (node == null)
+            {
+                return false;
+            }
+
+            if (node.Type != null)
+            {
+                if (node.Type.Kind == SyntaxKind.IdentifierName && ((IdentifierNameSyntax)node.Type).Identifier.IsVar && node.Deconstruction != null && !node.Deconstruction.OpenParenToken.IsMissing && !node.Deconstruction.CloseParenToken.IsMissing)
+                {
+                    // `var (..., ....)` with var and both parens present
+                    return true;
+                }
+
+                if (!topLevel && !node.Type.IsMissing && node.Variables.Count == 1 && !node.Variables[0].Identifier.IsMissing)
```

Comment:
`int x = M();` should not be parsed as a deconstruction-declaration, it is a regular local-declaration. So `int x` should only appear inside the deconstruction, not at the top-level.

On the other hand, the parsing already ensures that, so maybe that isn't needed for this check. I'll see if I can remove.